cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -2758,6 +2773,438 @@ struct GroupedOneFactory {
+struct GroupedListImpl<Type, enable_if_t<is_base_binary_type<Type>::value ||
+                                         std::is_same<Type, FixedSizeBinaryType>::value>>
+    final : public GroupedAggregator {
+  using Allocator = arrow::stl::allocator<char>;
+  using StringType = std::basic_string<char, std::char_traits<char>, Allocator>;
+  using GetSet = GroupedValueTraits<Type>;
+
+  Status Init(ExecContext* ctx, const std::vector<ValueDescr>&,
+              const FunctionOptions* options) override {
+    ctx_ = ctx;
+    allocator_ = Allocator(ctx_->memory_pool());
+    // out_type_ initialized by GroupedListInit
+    groups_ = TypedBufferBuilder<uint32_t>(ctx_->memory_pool());
+    values_bitmap_ = TypedBufferBuilder<bool>(ctx_->memory_pool());
+    return Status::OK();
+  }
+
+  Status Resize(int64_t new_num_groups) override {
+    num_groups_ = new_num_groups;
+    return Status::OK();
+  }
+
+  Status Consume(const ExecBatch& batch) override {
+    const auto values_array_data = batch[0].array();
+    int64_t num_values = values_array_data->length;
+    int64_t offset = values_array_data->offset;
+
+    const auto groups_array_data = batch[1].array();
+    DCHECK_EQ(groups_array_data->offset, 0);
+    const auto* groups = groups_array_data->GetValues<uint32_t>(1, 0);
+    RETURN_NOT_OK(groups_.Append(groups, num_values));
+
+    if (batch[0].null_count() == 0) {
+      RETURN_NOT_OK(values_bitmap_.Append(num_values, true));
+    } else {
+      const uint8_t* values_bitmap = values_array_data->buffers[0]->data();
+      RETURN_NOT_OK(GroupedValueTraits<BooleanType>::AppendBuffers(
+          &values_bitmap_, values_bitmap, offset, num_values));
+    }
+    return AppendBuffers<Type>(batch);
+  }
+
+  template <typename T = Type>
+  enable_if_t<is_base_binary_type<T>::value, Status> AppendBuffers(
+      const ExecBatch& batch) {
+    int64_t num_values = batch[0].array()->length;
+    num_args_ += num_values;
+    return VisitGroupedValues<Type>(
+        batch,
+        [&](uint32_t group, util::string_view val) -> Status {
+          values_.emplace_back(StringType(val.data(), val.size(), allocator_));
+          return Status::OK();
+        },
+        [&](uint32_t group) -> Status {
+          values_.emplace_back("");
+          return Status::OK();
+        });
+  }
+
+  template <typename T = Type>
+  enable_if_t<std::is_same<T, FixedSizeBinaryType>::value, Status> AppendBuffers(
+      const ExecBatch& batch) {
+    const auto values_array_data = batch[0].array();
+    int64_t num_values = values_array_data->length;
+
+    std::string str(
+        checked_cast<const FixedSizeBinaryType&>(*batch[0].type()).byte_width(), '0');
+    StringType s(str.data(), str.size(), allocator_);
+
+    //    values_.reserve(num_values * checked_cast<const
+    //    FixedSizeBinaryType&>(*batch[0].type()).byte_width());
+    values_.resize(values_.size() + num_values, s);
+    CopyDataUtils<FixedSizeBinaryType>::CopyData(
```

Comment:
I don't seem to be getting very far with `FixedBinaryType` and `CopyData`. How to reserve the correct amount of space in the `std::vector`? `resize` seems to overwrite the copied values sometimes. How to make `reserve` work for `util::optional<StringType>`? Do these offsets make sense?
https://github.com/apache/arrow/blob/ece0e23f1fd5a2ab403e4a645ed01a9c257c0260/cpp/src/arrow/compute/kernels/copy_data_internal.h#L70-L82

Reply:
Ah, I forgot it required you to preallocate data, hmm. In that case I might not bother with that, just note that you can use similar ideas to know how much you need to append in AppendBuffers, if that makes sense.